test/writeasync_test.dart

```diff
@@ -0,0 +1,244 @@
+  });
+
+  test('Realm.writeAsync returns valid objects', () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    final person = await realm.writeAsync(() {
+      return realm.add(Person('John'));
+    });
+
+    expect(person.name, 'John');
+  });
+
+  test('Realm.writeAsync throws user exception', () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    try {
+      await realm.writeAsync(() {
+        throw Exception('User exception');
+      });
+    } on Exception catch (e) {
+      expect(e.toString(), 'Exception: User exception');
+    }
+  });
+
+  test('Realm.writeAsync FIFO order ensured', () async {
+    final acquisitionOrder = <int>[];
+    final futures = <Future<void>>[];
+
+    final realm = getRealm(Configuration.local([Person.schema]));
+
+    for (var i = 0; i < 5; i++) {
+      futures.add(realm.writeAsync(() {
+        acquisitionOrder.add(i);
+      }));
+    }
+
+    await Future.wait(futures);
+
+    expect(acquisitionOrder, [0, 1, 2, 3, 4]);
+  });
+
+  test('Realm.beginWriteAsync with cancellation token', () async {
+    final realm1 = getRealm(Configuration.local([Person.schema]));
+    final realm2 = getRealm(Configuration.local([Person.schema]));
+    final t1 = realm1.beginWrite();
+
+    final token = CancellationToken();
+    Future<void>.delayed(Duration(milliseconds: 1)).then((value) => token.cancel());
+
+    await expectLater(() async => await realm2.beginWriteAsync(cancellationToken: token), throwsA(isA<CancelledException>()));
+
+    t1.rollback();
+  });
+
+  test('Realm.writeAsync with cancellation token', () async {
+    final realm1 = getRealm(Configuration.local([Person.schema]));
+    final realm2 = getRealm(Configuration.local([Person.schema]));
+    final t1 = realm1.beginWrite();
+
+    final token = CancellationToken();
+    Future<void>.delayed(Duration(milliseconds: 1)).then((value) => token.cancel());
+
+    await expectLater(() async => await realm2.writeAsync(() {}, cancellationToken: token), throwsA(isA<CancelledException>()));
+
+    t1.rollback();
+  });
+
+  test('Realm.beginWriteAsync when canceled after write lock obtained is a no-op', () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+
+    final token = CancellationToken();
+    final transaction = await realm.beginWriteAsync(cancellationToken: token);
+    token.cancel();
+
+    expect(transaction.isOpen, true);
+    expect(realm.isInTransaction, true);
+
+    transaction.rollback();
+  });
+
+  test('Realm.writeAsync when canceled after write lock obtained rolls it back', () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+
+    final token = CancellationToken();
+    await expectLater(
+        () => realm.writeAsync(() {
+              realm.add(Person('A'));
+              token.cancel();
+              realm.add(Person('B'));
+            }, cancellationToken: token),
+        throwsA(isA<CancelledException>()));
```

Comment:
```suggestion
    await expectLater(
        realm.writeAsync(() {
          realm.add(Person('A'));
          token.cancel();
          realm.add(Person('B'));
        }, cancellationToken: token),
        throwsA(isA<CancelledException>()));
```

Reply:
No, but this one does:

```dart
  test('example', () async {
    Future<void> test() {
      throw Exception('before first await!');
    }

    await expectLater(test(), throwsA(isA<Exception>()));
  });
```

Reply:
Yes, but that is just [bad style](https://dart.dev/guides/language/effective-dart/usage#prefer-asyncawait-over-using-raw-futures), and that is not how `writeAsync` is written?

Reply:
Anyway, I'm fine with both 👍

Reply:
> How would you return a future from test

Not sure I understand - in the example above, we're refactoring the implementation of `writeAsync`, nothing to do with tests. Yes, it's returning a raw future, but it's not using `.then/.catch` which is "bad style" the way I read the guide. In any case, I still prefer the lambda style, because it's consistent with non-promise asserts:

```dart
expect(() => doSomething(), throws<Exception>())
await expectLater(() => doSomethingAsync(), throws<Exception>());
```

Reply:
As I read the guide you should use `async` unless dropping it doesn't change behaviour.

You can write:
```dart
Future<T> writeAsync(...) {
  return _writeAsyncInternal(...);
}
```
but replacing:
```dart
Future<T> writeAsync(...) async {
  if (something) {
    throw Exception();
  }

  return awaIt _writeAsyncInternal(...);
}
```
with
```dart
Future<T> writeAsync(...) {
  if (something) {
    throw Exception();
  }

  return _writeAsyncInternal(...);
}
```
does change behaviour and should be avoided.

Anyway, as I said, I can live with the defensive code in the tests, as long as we don't actually ever push such behaviour onto our users.

Reply:
I changed the code to assert on the future directly rather than a lambda.